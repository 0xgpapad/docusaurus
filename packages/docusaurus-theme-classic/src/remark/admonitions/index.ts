/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

/* eslint-disable */

import type {Transformer} from 'unified';
import visit from 'unist-util-visit';

const NEWLINE = '\n';

const config = {
  tag: ':::',
  keywords: [
    'secondary',
    'info',
    'success',
    'danger',
    'note',
    'tip',
    'warning',
    'important',
    'caution',
  ],
};

function escapeRegExp(s: string): string {
  return s.replace(new RegExp(`[-[\\]{}()*+?.\\\\^$|/]`, 'g'), '\\$&');
}

export default function plugin(this: any): Transformer {
  const keywords = Object.values(config.keywords).map(escapeRegExp).join('|');
  const tag = escapeRegExp(config.tag);
  const regex = new RegExp(`${tag}(${keywords})(?: *(.*))?\n`);
  const escapeTag = new RegExp(escapeRegExp(`\\${config.tag}`), 'g');

  // the tokenizer is called on blocks to determine if there is an admonition present and create tags for it
  function blockTokenizer(this: any, eat: any, value: string, silent: boolean) {
    // stop if no match or match does not start at beginning of line
    const match = regex.exec(value);
    if (!match || match.index !== 0) return false;
    // if silent return the match
    if (silent) return true;

    const now = eat.now();
    const [opening, keyword, title] = match;
    const food = [];
    const content = [];

    // consume lines until a closing tag
    let idx = 0;
    while ((idx = value.indexOf(NEWLINE)) !== -1) {
      // grab this line and eat it
      const next = value.indexOf(NEWLINE, idx + 1);
      const line =
        next !== -1 ? value.slice(idx + 1, next) : value.slice(idx + 1);
      food.push(line);
      value = value.slice(idx + 1);
      // the closing tag is NOT part of the content
      if (line.startsWith(config.tag)) break;
      content.push(line);
    }

    // consume the processed tag and replace escape sequences
    const contentString = content.join(NEWLINE).replace(escapeTag, config.tag);
    const add = eat(opening + food.join(NEWLINE));

    // parse the content in block mode
    const exit = this.enterBlock();
    const contentNodes = this.tokenizeBlock(contentString, now);
    exit();

    const element = {
      type: 'admonitionHTML',
      data: {
        hName: 'admonition',
        hProperties: {
          title,
          type: keyword,
        },
      },
      children: contentNodes,
    };

    return add(element);
  }

  // add tokenizer to parser after fenced code blocks
  const Parser = (this as any).Parser.prototype;
  Parser.blockTokenizers.admonition = blockTokenizer;
  Parser.blockMethods.splice(
    Parser.blockMethods.indexOf('fencedCode') + 1,
    0,
    'admonition',
  );
  Parser.interruptParagraph.splice(
    Parser.interruptParagraph.indexOf('fencedCode') + 1,
    0,
    ['admonition'],
  );
  Parser.interruptList.splice(
    Parser.interruptList.indexOf('fencedCode') + 1,
    0,
    ['admonition'],
  );
  Parser.interruptBlockquote.splice(
    Parser.interruptBlockquote.indexOf('fencedCode') + 1,
    0,
    ['admonition'],
  );

  return function transformer(tree) {
    // escape everything except admonitionHTML nodes
    // @ts-ignore FIXME
    visit(
      tree,
      (node: any) => {
        return node.type !== 'admonitionHTML';
      },
      function visitor(node: any) {
        if (node.value) {
          node.value = node.value.replace(escapeTag, config.tag);
        }
        return node;
      },
    );
  };
}
