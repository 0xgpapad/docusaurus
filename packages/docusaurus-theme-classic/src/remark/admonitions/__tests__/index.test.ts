/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import path from 'path';
import remark from 'remark';
import vfile from 'to-vfile';
import plugin from '../index';
import remark2rehype from 'remark-rehype';
import stringify from 'rehype-stringify';

const processFixture = async (name) => {
  const filePath = path.join(__dirname, '__fixtures__', `${name}.md`);
  const file = await vfile.read(filePath);

  const result = await remark()
    .use(plugin)
    .use(remark2rehype)
    .use(stringify)
    .process(file);

  return result.toString();
};

describe('admonitions remark plugin', () => {
  it('base', async () => {
    const result = await processFixture('base');
    expect(result).toMatchSnapshot();
  });
});
