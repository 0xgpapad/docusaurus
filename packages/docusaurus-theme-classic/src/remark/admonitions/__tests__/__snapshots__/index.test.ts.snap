// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`admonitions remark plugin base 1`] = `
"<p>The blog feature enables you to deploy in no time a full-featured blog.</p>
<admonition title="Sample Title" type="info"><p>Check the <a href="./api/plugins/plugin-content-blog.md">Blog Plugin API Reference documentation</a> for an exhaustive list of options.</p></admonition>
<h2>Initial setup {#initial-setup}</h2>
<p>To set up your site's blog, start by creating a <code>blog</code> directory.</p>
<admonition type="tip"><p>Use the <strong><a href="introduction.md#fast-track">Fast Track</a></strong> to understand Docusaurus in <strong>5 minutes ‚è±</strong>!</p><p>Use <strong><a href="https://docusaurus.new">docusaurus.new</a></strong> to test Docusaurus immediately in your browser!</p></admonition>"
`;
